### PR TITLE
feat(workers): add snowflake-sync example

### PR DESCRIPTION
**Pull request by atvaccaro**
## What

Adds `examples/workers/syncs/snowflake-sync/` — a complete, runnable warehouse→Notion sync. It reads a configurable Snowflake query (`SNOWFLAKE_SYNC_QUERY`) using the same key-pair auth as the `snowflake-query` tool, and syncs the rows into a **managed** Notion database via `worker.database` + `worker.sync`. Paginated with `LIMIT`/`OFFSET` (200/page), `mode: "replace"`.

`schema.ts` and `transform.ts` are a worked example (columns `id, name, email, status, updated_at`) with prominent "edit to match your query's columns" guidance. Runnable once `SNOWFLAKE_*` env is set; the README notes the real read boundary is the Snowflake role's grants (use a read-only role) and points at incremental-mode + cursor as a follow-up for large/changing tables.

## Verification

- `npm run check`, `npm run build`, `npm test` (19 offline tests: the `LIMIT`/`OFFSET` page-SQL builder + row→property mapping, incl. Snowflake UPPERCASE keys and null-date handling), `prettier --check`, `markdownlint` — all clean
- Deploy-smoke: deployed to dev; `sync snowflakeSync` discovered (`snowflake-sdk` bundles fine; env is validated lazily so discovery works without creds). A full run requires `SNOWFLAKE_*` credentials.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by atvaccaro** [comment]
* **#31** <a href="https://app.graphite.com/github/pr/makenotion/notion-cookbook/31?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/makenotion/notion-cookbook/31?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#30** <a href="https://app.graphite.com/github/pr/makenotion/notion-cookbook/30?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#29** <a href="https://app.graphite.com/github/pr/makenotion/notion-cookbook/29?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#28** <a href="https://app.graphite.com/github/pr/makenotion/notion-cookbook/28?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#27** <a href="https://app.graphite.com/github/pr/makenotion/notion-cookbook/27?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26** <a href="https://app.graphite.com/github/pr/makenotion/notion-cookbook/26?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25** <a href="https://app.graphite.com/github/pr/makenotion/notion-cookbook/25?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24** <a href="https://app.graphite.com/github/pr/makenotion/notion-cookbook/24?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: examples/workers/syncs/snowflake-sync/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+const database = worker.database("rows", {
+  type: "managed",
+  initialTitle: INITIAL_TITLE,
+  primaryKeyProperty: PRIMARY_KEY,
+  schema: rowSchema,
+})
+
+worker.sync("snowflakeSync", {
+  database,
+  mode: "replace",
```
Comment (ksinder) [inline examples/workers/syncs/snowflake-sync/src/index.ts:28]: **[Non-blocking] Replace-mode sweep can delete live rows when OFFSET paging is not a stable snapshot**

Medium. In replace mode the platform treats every row emitted across the whole cycle as the full dataset and archives any existing Notion row whose key was not seen. `syncs.mdx` states it plainly: "any rows not seen during that cycle are automatically deleted." This worker pages with `LIMIT`/`OFFSET`, and each `fetchPage` opens a fresh connection and runs an independent query, so there is no consistent snapshot across pages. Two ways a live row gets dropped and then deleted from Notion:

- No `ORDER BY`: Snowflake may return rows in a different order on each page query, so a boundary row can land in zero pages.
- The source table changes mid-cycle: a delete in an earlier page shifts every later `OFFSET` by one and skips a row.

Either way the skipped row is absent from the cycle, mark-and-sweep archives it, and it can reappear next run (data flapping plus transient deletion of live records). For reference code, make `ORDER BY` on a stable, immutable key a hard requirement rather than a suggestion, and note that OFFSET paging over a mutating table is not snapshot-consistent. See the related README note.

File: examples/workers/syncs/snowflake-sync/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+**`src/transform.ts`** — maps a raw Snowflake row to a sync change. Column
+lookups use `row["COLUMN"] ?? row["column"]` to handle Snowflake's UPPERCASE
+default and explicit lowercase aliases. Add or remove `Builder.*` calls to
+match your schema.
+
+## Pagination
+
+The worker uses `LIMIT 200 OFFSET N` to page through results. The platform
+loops execute() calls until `hasMore` is false. For queries returning millions
+of rows this is reliable but sequential. An `ORDER BY` clause in
```
Comment (ksinder) [inline examples/workers/syncs/snowflake-sync/README.md:120]: **[Non-blocking] Make the ordering requirement firm and reconcile the row-count guidance**

Low/Medium. Given the replace-mode sweep behavior (see my note on `src/index.ts`), `ORDER BY <stable key>` should be required, not "recommended." Separately, this section pitches `LIMIT`/`OFFSET` as "reliable but sequential" for "queries returning millions of rows," but the platform guide scopes replace mode to "smaller datasets (under 10k records)" and steers large or frequently-changing tables to incremental + cursor. Suggest aligning this paragraph with that guidance so a reader does not run a million-row replace sync and hit skip-and-delete churn.

File: examples/workers/syncs/snowflake-sync/src/snowflake.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      Object.entries(value).map(([key, nested]) => [
+        key,
+        normalizeValue(nested),
+      ])
+    )
+  }
+  return String(value)
+}
+
+export function clamp(
```
Comment (ksinder) [inline examples/workers/syncs/snowflake-sync/src/snowflake.ts:94]: **[Non-blocking] `clamp` and `readPositiveInt` are exported but never used**

Low. Neither helper is called anywhere in this example; `PAGE_SIZE` is a hardcoded `200` in `index.ts`. They are leftovers from `snowflake-query/src/snowflake.ts`, where the same two functions are private and feed the row cap. For reference code, either wire them up (for example `const PAGE_SIZE = clamp(readPositiveInt("SNOWFLAKE_SYNC_PAGE_SIZE", 200), 200, 1000)` plus a line in `.env.example`) or delete both. As written they read as accidental, untested API surface.

File: examples/workers/syncs/snowflake-sync/src/snowflake.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+ * @param limit  Maximum rows to return for this page.
+ * @param offset Row offset to start from.
+ */
+export async function fetchPage(
+  query: string,
+  limit: number,
+  offset: number
+): Promise<Record<string, unknown>[]> {
+  const sql = buildPageSql(query, limit, offset)
+  const conn = createConnection()
```
Comment (ksinder) [inline examples/workers/syncs/snowflake-sync/src/snowflake.ts:141]: **[Non-blocking] A fresh authenticated connection per page**

Low. `fetchPage` creates a connection, does a JWT handshake, and destroys it on every page. The `snowflake-query` tool does the same, but it runs one query per invocation; here it repeats for every page of a full sync, so a large table means many connect/auth/teardown cycles within a single sync cycle. Consider holding one connection for the duration of the cycle, or using `snowflake-sdk`'s streaming cursor so the server paginates over a single statement. Not a correctness issue, just efficiency in code others will copy.

File: examples/workers/syncs/snowflake-sync/README.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+[Adapting to your query](#adapting-to-your-query) to change it.
+
+## Prerequisites
+
+- Node >= 22, npm >= 10.9.2
+- A Snowflake account with a read-only service role and an RSA key pair for
+  JWT authentication. See the
+  [snowflake-query README](../../tools/snowflake-query/README.md) for key-pair
+  setup instructions.
+- The `ntn` CLI installed and authenticated (`ntn auth login`).
```
Comment (ksinder) [inline examples/workers/syncs/snowflake-sync/README.md:28]: **[Non-blocking] `ntn auth login` differs from every other example**

Low. The prereqs say `ntn auth login`, but all ten other example READMEs in the repo (including the sibling duckdb-sync-demo in this same stack) use `ntn login`. Looks like a typo; align it so copy-paste works.

**Review by ksinder — APPROVED** [review]
Reviewed snowflake-sync against the workers sync guide (`syncs.mdx`), the `@notionhq/workers` SDK reference, and the existing `snowflake-query` tool it borrows from. The offline tests are good: `buildPageSql` and `rowToChange` cover the real edge cases (trailing semicolons, UPPERCASE vs lowercase Snowflake keys, null/missing/empty ID, null dates). CI is green. Approving.

One finding is worth real attention. It is non-blocking but matters for an example people copy: replace-mode mark-and-sweep combined with OFFSET pagination can delete live rows. Full explanation inline on `src/index.ts` and the README pagination section. The rest are small cleanups.

Verified against ground truth:
- `syncs.mdx` confirms replace mode deletes any row "not seen during that cycle" and is "best for smaller datasets (under 10k records)." That is the basis for the data-loss finding.
- `clamp` and `readPositiveInt` are unused in this example; in the `snowflake-query` tool the same two helpers are private and actually feed the row cap. They are copy-paste leftovers here.
- `ntn auth login` in the prereqs differs from all ten other example READMEs in the repo (including the sibling duckdb-sync-demo), which use `ntn login`.

[review written by Claude Code on behalf of @ksinder]

**Comment by atvaccaro** [comment]
### Merge activity

* **Jun 26, 9:05 PM UTC**: @atvaccaro merged this pull request with [Graphite](https://app.graphite.com/github/pr/makenotion/notion-cookbook/31).
